### PR TITLE
Combines block records in single dict that includes all associated wfs

### LINKED ISSUE #11244: RucioInjector deleting blocks prematurely due to multiple workflows writting on them
**Impact of the bug**
T0 Agent

**Describe the bug**
Current deleteBlocks logic is faulty. In order to delete a block, the component should check that there are no workflows that need the files in the block. Nonetheless, right now it assumes each clock is associated with a single workflow. If only one of the workflows is deletable, RucioInjector deletes the block, even to other workflows may still need it.

Full description of the error can be found here:
https://its.cern.ch/jira/browse/CMSTZ-1001

**How to reproduce it**
Running a replay with 2 runs that share some datasets. You must stop prompt reconstruction workflows for the seconde run. Set DeletionDelay to a short value. After the first run finishes processing, you will see that several blocks will be deleted, including those that are still needed by the second run.

**Expected behavior**
If a block is needed for processing, it should not be deleted

### PR DESCRIPTION
**Pull request by germanfgv**
Fixes #11244 

#### Status
Ready

#### Description
Turn list of blocks into a dictionary to avoid duplicate registers for single block. Turns 'workflowname' into a set of workflows that generated the block

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
None

#### External dependencies / deployment changes
None

### CODE REVIEW

**Comment by amaltaro** [comment]
@germanfgv I am not sure I fully undertand the problem you reported. However, looking into this line that you implemented:
```
if block['workflowNames'].issubset(deletableWfsDict)
```

it makes me think that, regardless whether we have your changes in place or not, `block['workflowNames']` will be a subset of `deletableWfsDict` if it has only 1 workflow or multiple workflows in the set.

Unless completedWorkflows is always fully built before deletableWorkflows, then in this case yes, it would only evaluate to True once all the completedWorkflows are also deletableWorkflows.

**Comment by germanfgv** [comment]
> @germanfgv I am not sure I fully undertand the problem you reported. However, looking into this line that you implemented:
> 
> ```
> if block['workflowNames'].issubset(deletableWfsDict)
> ```
> 
> it makes me think that, regardless whether we have your changes in place or not, `block['workflowNames']` will be a subset of `deletableWfsDict` if it has only 1 workflow or multiple workflows in the set.
> 
> Unless completedWorkflows is always fully built before deletableWorkflows, then in this case yes, it would only evaluate to True once all the completedWorkflows are also deletableWorkflows.

Maybe this concern is address with my second commit. Before, we were disregarding incomplete workflows in the query due to `HAVING COUNT(*) = SUM(dbsbuffer_workflow.completed)`.

Because of this, you are right that `if block['workflowNames'].issubset(deletableWfsDict)` would almost always evaluate true (_Almost_ because _deletable_ and _completed_ are not the same. E.g. a completed repack workflow that cannont be deleted because its output is being used by a promptReco workflow)

_deletable_ > _completed_. By removing the completion check in the query in https://github.com/dmwm/WMCore/pull/11245/commits/0bf4285a781e7a35694ae59355765fca435747a4, now we can properly avoid deleting blocks that belong to incomplete workflows

File: src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py [hunk 1]
```diff
@@ -86,21 +85,29 @@ def format(self, result):
...
-                    record[pyKey] = set()
-                    record[pyKey].add(sites)
-                else:
-                    record[pyKey] = record.pop(dbKey)
+        listResults = DBFormatter.formatDict(self, result)
+        dictResults = {}
+        for record in listResults:
+            # Populates results dict and adds all workflows of the same block to a single record
+            blockname = record['blockname']
+            if blockname in dictResults:
```
Comment (todor-ivanov) [inline src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py:99]: In general the DAO must return the list of results as it is from the DABASE and no extra data reformatting should be needed. It seems to me that if this extra aggregation by block name here in the python code is needed, then something is not quite working as we expect in the `GROUP BY` clause in the SQL query.

File: src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py [hunk 2]
```diff
@@ -86,21 +85,29 @@ def format(self, result):
...
+                dictResults[blockname]['workflowNames'].add(record['name'])
+                dictResults[blockname]['sites'].add(record['site'])
```
Comment (todor-ivanov) [inline src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py:101]: IIUC this line here would just add the site to the set of sites of an already existing block name, skipping the values  of the rest of the columns in the record. I am not sure this is safe. Are we not losing information here?

File: src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py [hunk 3]
```diff
@@ -86,21 +85,29 @@ def format(self, result):
         # NOTE: We need to rename all the keys to follow the cammelCase standard. And also to comply
```
Comment (todor-ivanov) [inline src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py:85]: If we decide to return a dictionary instead of a list of records (for which I am not 100% sure yet), all that needs to be reflected in the doc string as well.

File: src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py [hunk 4]
```diff
@@ -86,21 +85,29 @@ def format(self, result):
...
-                    record[pyKey].add(sites)
-                else:
-                    record[pyKey] = record.pop(dbKey)
+        listResults = DBFormatter.formatDict(self, result)
+        dictResults = {}
+        for record in listResults:
+            # Populates results dict and adds all workflows of the same block to a single record
+            blockname = record['blockname']
+            if blockname in dictResults:
+                dictResults[blockname]['workflowNames'].add(record['name'])
```
Comment (todor-ivanov) [inline src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py:100]: same comment as bellow.

**Review by todor-ivanov — CHANGES_REQUESTED** [review]
Hi @germanfgv Thanks for the suggested change. 
I am still having some concerns about this extra aggregation that is needed in the python code. This is all a sign that we are ignoring the uniqueness of the records as they come from the `CROUP BY` SQL statement, which to me seems to be dangerous. 
In parallel to the already existing changes, which as far as I understood are already working in one of the T0 machines in production, I'd like to be able to test few of those SQL queries and DAO outcomes myself in one of the machines affected by this bug, if you do not mind.

File: src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py [hunk 5]
```diff
@@ -49,7 +49,6 @@ class GetCompletedBlocks(DBFormatter):
                       dbsbuffer_dataset_subscription.site,
                       dbsbuffer_workflow.name,
                       dbsbuffer_block.create_time
-             HAVING COUNT(*) = SUM(dbsbuffer_workflow.completed)
```
[the remark this answers is not in the document]
Comment (germanfgv) [reply src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py:52]: That was useful in [GetDeletableBlocks.py](https://github.com/dmwm/WMCore/blob/master/src/python/WMComponent/RucioInjector/Database/MySQL/GetDeletableBlocks.py) because that query was not grouping by workflow name. Here, if you exclude records with workflows that are not complete, you will still have other records with other workflows associated to the block, so the block will still end up in the results, even with uncompleted associated workflows.
 
Even worst, as we are excluding the uncompleted workflows from the query, later we are not taking thos uncompleted workflows into account when we compare against the `deletableWfsDict` list

File: src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py (hunk 1 above)
Comment (germanfgv) [reply src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py:99]: We can do this aggregation outside of the DAO. Also, we may be able to modify the query so it ignores blocks with blocks that are not completed, but then that's probably redundant with the deletableWorkflows query

File: src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py (hunk 2 above)
Comment (germanfgv) [reply src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py:101]: the other columns are block name, location, dataset, and creation time. All of those are the same for each block, regardless of the subscription in the record.

File: src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py (hunk 3 above)
Comment (germanfgv) [reply src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py:85]: Agreed. Let's decide what we will do first and then I can make the changes

File: src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (todor-ivanov) [reply src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py:52]: >  because that query was not grouping by workflow name
That is correct. 
I have double checked now, that introducing the `dbsbuffer_workflow.name` in the `SELECT`  and `GROUP BY` conditions splits the grouping further down, creating  multiple groups per block - one for every associated workflow. And then it does count  the completed workflows but they enter the  query every time a workflow completes..... so if we have 1 block with 3 associated PromptReco requests out of which only two are completed, we will have that block returned 2 times  in the output. Once the third Prompt Reco is complete the same block would enter the output result 3 times.

This makes the rest of the logic quite hard to follow indeed.

[the remark this answers is not in the document]
Comment (todor-ivanov) [reply src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py:52]: And since it is quite subtle detail and difficult to understand and remember. Let me put some concrete examples:

Here is the output from the current SQL query with the counts  per group included in the output (filtered for a single block):
```
SQL> SELECT
  2     count(*),
  3     dbsbuffer_block.blockname,
  4     dbsbuffer_location.pnn,
  5     dbsbuffer_dataset.path,
  6     dbsbuffer_dataset_subscription.site,
  7     dbsbuffer_workflow.name,
  8     dbsbuffer_block.create_time
  9  FROM dbsbuffer_dataset_subscription
 10  INNER JOIN dbsbuffer_dataset ON
 11     dbsbuffer_dataset.id = dbsbuffer_dataset_subscription.dataset_id
 12  INNER JOIN dbsbuffer_block ON
 13     dbsbuffer_block.dataset_id = dbsbuffer_dataset_subscription.dataset_id
 14  INNER JOIN dbsbuffer_file ON
 15     dbsbuffer_file.block_id = dbsbuffer_block.id
 16  INNER JOIN dbsbuffer_workflow ON
 17     dbsbuffer_workflow.id = dbsbuffer_file.workflow
 18  INNER JOIN dbsbuffer_location ON
 19     dbsbuffer_location.id = dbsbuffer_block.location
 20  WHERE dbsbuffer_dataset_subscription.delete_blocks = 1
 21  AND dbsbuffer_dataset_subscription.subscribed = 1
 22  AND dbsbuffer_block.status = 'Closed'
 23  AND dbsbuffer_block.deleted = 0
 24  AND dbsbuffer_block.blockname = '/ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00'
 25   GROUP BY dbsbuffer_block.blockname,
 26     dbsbuffer_location.pnn,
 27     dbsbuffer_dataset.path,
 28     dbsbuffer_dataset_subscription.site,
 29     dbsbuffer_workflow.name,
 30     dbsbuffer_block.create_time
 31  HAVING COUNT(*) = SUM(dbsbuffer_workflow.completed)
 32  ORDER BY dbsbuffer_workflow.name;

  COUNT(*) BLOCKNAME														    PNN 	    PATH				     SITE		NAME					 CREATE_TIME
---------- ------------------------------------------------------------------------------------------------------------------------ --------------- ---------------------------------------- ------------------ ---------------------------------------- -----------
	82 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_Disk	PromptReco_Run357080_ZeroBias		  1660249466
	82 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_MSS	PromptReco_Run357080_ZeroBias		  1660249466

``` 

Now lets release the `HAVING COUNT(*) = SUM(dbsbuffer_workflow.competed)` constraint:

```
SQL> SELECT
  2     count(*),
  3     dbsbuffer_block.blockname,
  4     dbsbuffer_location.pnn,
  5     dbsbuffer_dataset.path,
  6     dbsbuffer_dataset_subscription.site,
  7     dbsbuffer_workflow.name,
  8     dbsbuffer_block.create_time
  9  FROM dbsbuffer_dataset_subscription
 10  INNER JOIN dbsbuffer_dataset ON
 11     dbsbuffer_dataset.id = dbsbuffer_dataset_subscription.dataset_id
 12  INNER JOIN dbsbuffer_block ON
 13     dbsbuffer_block.dataset_id = dbsbuffer_dataset_subscription.dataset_id
 14  INNER JOIN dbsbuffer_file ON
 15     dbsbuffer_file.block_id = dbsbuffer_block.id
 16  INNER JOIN dbsbuffer_workflow ON
 17     dbsbuffer_workflow.id = dbsbuffer_file.workflow
 18  INNER JOIN dbsbuffer_location ON
 19     dbsbuffer_location.id = dbsbuffer_block.location
 20  WHERE dbsbuffer_dataset_subscription.delete_blocks = 1
 21  AND dbsbuffer_dataset_subscription.subscribed = 1
 22  AND dbsbuffer_block.status = 'Closed'
 23  AND dbsbuffer_block.deleted = 0
 24  AND dbsbuffer_block.blockname = '/ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00'
 25   GROUP BY dbsbuffer_block.blockname,
 26     dbsbuffer_location.pnn,
 27     dbsbuffer_dataset.path,
 28     dbsbuffer_dataset_subscription.site,
 29     dbsbuffer_workflow.name,
 30     dbsbuffer_block.create_time
 31  ORDER BY dbsbuffer_workflow.name;

  COUNT(*) BLOCKNAME														    PNN 	    PATH				     SITE		NAME					 CREATE_TIME
---------- ------------------------------------------------------------------------------------------------------------------------ --------------- ---------------------------------------- ------------------ ---------------------------------------- -----------
	82 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_Disk	PromptReco_Run357080_ZeroBias		  1660249466
	82 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_MSS	PromptReco_Run357080_ZeroBias		  1660249466
	77 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_Disk	PromptReco_Run357081_ZeroBias		  1660249466
	77 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_MSS	PromptReco_Run357081_ZeroBias		  1660249466
	 2 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_Disk	PromptReco_Run357096_ZeroBias		  1660249466
	 2 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_MSS	PromptReco_Run357096_ZeroBias		  1660249466
	 8 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_Disk	PromptReco_Run357098_ZeroBias		  1660249466
	 8 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_MSS	PromptReco_Run357098_ZeroBias		  1660249466
	 7 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_Disk	PromptReco_Run357100_ZeroBias		  1660249466
	 7 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_MSS	PromptReco_Run357100_ZeroBias		  1660249466
	13 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_Disk	PromptReco_Run357101_ZeroBias		  1660249466

	13 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_MSS	PromptReco_Run357101_ZeroBias		  1660249466
	 9 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_Disk	PromptReco_Run357102_ZeroBias		  1660249466
	 9 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_MSS	PromptReco_Run357102_ZeroBias		  1660249466
	 1 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_Disk	PromptReco_Run357106_ZeroBias		  1660249466
	 1 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_MSS	PromptReco_Run357106_ZeroBias		  1660249466
	 2 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_Disk	PromptReco_Run357112_ZeroBias		  1660249466
	 2 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_MSS	PromptReco_Run357112_ZeroBias		  1660249466

18 rows selected.

```

It is obvious that  this block would continue entering the output of the original query every time a PromptReco completes until the result of the original query does not give exactly 18 records as the one above.

And indeed the expected behavior is achieved once we have the field `dbsbuffer_workflow.name` , which is further splitting the groups, removed.

Here is the query without the extra splitting and without the constraint for all the associated workflows to be completed as before:
```
SQL> SELECT
  2     count(*),
  3     dbsbuffer_block.blockname,
  4     dbsbuffer_location.pnn,
  5     dbsbuffer_dataset.path,
  6     dbsbuffer_dataset_subscription.site,
  7     dbsbuffer_block.create_time
  8  FROM dbsbuffer_dataset_subscription
  9  INNER JOIN dbsbuffer_dataset ON
 10     dbsbuffer_dataset.id = dbsbuffer_dataset_subscription.dataset_id
 11  INNER JOIN dbsbuffer_block ON
 12     dbsbuffer_block.dataset_id = dbsbuffer_dataset_subscription.dataset_id
 13  INNER JOIN dbsbuffer_file ON
 14     dbsbuffer_file.block_id = dbsbuffer_block.id
 15  INNER JOIN dbsbuffer_workflow ON
 16     dbsbuffer_workflow.id = dbsbuffer_file.workflow
 17  INNER JOIN dbsbuffer_location ON
 18     dbsbuffer_location.id = dbsbuffer_block.location
 19  WHERE dbsbuffer_dataset_subscription.delete_blocks = 1
 20  AND dbsbuffer_dataset_subscription.subscribed = 1
 21  AND dbsbuffer_block.status = 'Closed'
 22  AND dbsbuffer_block.deleted = 0
 23  AND dbsbuffer_block.blockname = '/ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00'
 24   GROUP BY dbsbuffer_block.blockname,
 25     dbsbuffer_location.pnn,
 26     dbsbuffer_dataset.path,
 27     dbsbuffer_dataset_subscription.site,
 28     dbsbuffer_block.create_time ;

  COUNT(*) BLOCKNAME														    PNN 	    PATH				     SITE		CREATE_TIME
---------- ------------------------------------------------------------------------------------------------------------------------ --------------- ---------------------------------------- ------------------ -----------
       201 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_Disk	 1660249466
       201 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_MSS	 1660249466

```

And here is the result if we apply the constraint for workflow completion.  
```
SQL> SELECT
  2     count(*),
  3     dbsbuffer_block.blockname,
  4     dbsbuffer_location.pnn,
  5     dbsbuffer_dataset.path,
  6     dbsbuffer_dataset_subscription.site,
  7     dbsbuffer_block.create_time
  8  FROM dbsbuffer_dataset_subscription
  9  INNER JOIN dbsbuffer_dataset ON
 10     dbsbuffer_dataset.id = dbsbuffer_dataset_subscription.dataset_id
 11  INNER JOIN dbsbuffer_block ON
 12     dbsbuffer_block.dataset_id = dbsbuffer_dataset_subscription.dataset_id
 13  INNER JOIN dbsbuffer_file ON
 14     dbsbuffer_file.block_id = dbsbuffer_block.id
 15  INNER JOIN dbsbuffer_workflow ON
 16     dbsbuffer_workflow.id = dbsbuffer_file.workflow
 17  INNER JOIN dbsbuffer_location ON
 18     dbsbuffer_location.id = dbsbuffer_block.location
 19  WHERE dbsbuffer_dataset_subscription.delete_blocks = 1
 20  AND dbsbuffer_dataset_subscription.subscribed = 1
 21  AND dbsbuffer_block.status = 'Closed'
 22  AND dbsbuffer_block.deleted = 0
 23  AND dbsbuffer_block.blockname = '/ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00'
 24   GROUP BY dbsbuffer_block.blockname,
 25     dbsbuffer_location.pnn,
 26     dbsbuffer_dataset.path,
 27     dbsbuffer_dataset_subscription.site,
 28     dbsbuffer_block.create_time
 29  HAVING COUNT(*) = SUM(dbsbuffer_workflow.completed) ;

no rows selected
```   

It returns no rows as expected, because there are still workflows associated with this block which are not  completed.

[the remark this answers is not in the document]
Comment (todor-ivanov) [reply src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py:52]: I am afraid this same faulty logic applies for the extra granularity added by the multiple sites as  well as with the multiple associated workflows

File: src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py (hunk 1 above)
Comment (todor-ivanov) [reply src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py:99]: Thanks @germanfgv!  We largely discussed the problem an tested multiple times with you, and we now know in details what is going on. I'd suggest  we  continue as it is. And also since this DAO is used at a single place in the WMCore code, my previous remark that we should avoid the aggregation in the DAO and leave it at the component would only make the code at the component more unreadable. So lets proceed with what we already have. Just put the comments, as we discussed during our meeting, mentioning what are the background condition we rely on in order for that logic to work. So that we do remember those little but extremely important details. Thanks!

File: src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py (hunk 2 above)
Comment (todor-ivanov) [reply src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py:101]: As above, put this little detail at a comment in the code, please.

File: src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py (hunk 3 above)
Comment (todor-ivanov) [reply src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py:85]: Upon our discussion and tests, decision has been made about that - Lets return a dictionary and simply reflect that change in the DAO's docstring.

File: src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py [hunk 6]
```diff
@@ -58,28 +58,35 @@ def format(self, result):
         Format the query results into the proper dictionary expected at the upper layer Python code.
         The input should be a list of database objects each representing a line returned from the database
         with key names matching the column names from the sql query
-        The result should be a list of dictionaries one record per line returned from the database
-        with key names mapped to the python code expected structures.
+        The result should be a list of dictionaries one record per block returned from the database
```
Comment (todor-ivanov) [inline src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py:61]: The result is a dictionary of dictionaries, and having the block names as kyes. But that's minor.

**Review by todor-ivanov — APPROVED** [review]
Thanks @germanfgv , please squash those commits in a single one so we can merge it.

File: src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py (hunk 6 above)
Comment (vkuznet) [reply src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py:61]: Did you benchmark it? Having dict of dicts with dynamic keys usually leads to different problems. For instance, in order to parse it where all keys are unknown a-priory the parser code should go through all nested structs. The nested structures usually means that no streaming can be done which again is architectural choice but you should understand the consequences of such choice for your application. This may lead to large memory footprint in application level. By benchmark I mean try to create largest object with this data struct and write code which will parse it. Measure time, and CPU/RAM usage to have clear idea that usage of such data structure will not surprise you in production application.

Comment (todor-ivanov) [reply src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py:61]: Hi @vkuznet Thanks for mentioning that. This was exactly the reason why at the beginning I opposed the aproach in which the DAO is to do these aggregations in a dictionary form. I just did not expand on the reasons why. And thanks for pointing it out here. These are indeed valid concerns, but we had to compromise. The dynamic keys are needed because otherwise the whole logic for how those SQL queries are constructed and how the checks for deletable blocks are preformed need to be changed significantly. Which would be extremely dangerous now as we are in the middle of the data taking. 
Benchmark and scalability tests and memory footprint comparisons between the two DAO outputs are not done - this I am sure about. But we also very well know that no additional parsing of the returned object is done rather than a single iteration and a set comparison for all of its elements is happening. Which should not increase the code complexity more than to a O(n).  As of memory foot print, even if a single DAO output does increase in size due to the data structure change....well this one is used only once during the polling cycle and immediately gets out of scope after the method is executed, so again I wouldn't expect more than a constant increase in that regard either. And here are the reasons why I did agree with the so suggested change of the data structure returned by the DAO. 

And just a hint - the code is already applied through a patch in production so we are feeling quite confident it is working at least well enough for the current load at T0.

Comment (vkuznet) [reply src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py:61]: Todor, with all my respect, this does not answer my questions. I hope you also understand that by providing these  "explanations" we only hide the issue or delegate it to another part of infrastructure. When you claim that code in production and it runs, does it mean that you measure the impact of this change? What is it? Does this change RAM by 1% or double it. We may still run within the resource allocation but we may enter dangerous zone or thing may break if object will get larger. Let's be professional here. Please estimate the impact and provide concrete numbers. Without them any explanation does not make any sense to me but its impact may not be immediate but still valid and dangerous, especially during data taking.

Comment (germanfgv) [reply src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py:61]: While not a proper benchmark, I took a few points of data on the performance before and after the patch. These data points are taken with 9796 blocks in the system, a normal number after a few weeks of heavy data taking (we peaked at around 13k blocks during July). 

**Without patch**
These 9796 blocks translates into a list of 40665 elements. 
PSS increased around 50MB after calling the query
RucioInjectorPoller took around 30-40 secs to execute

**With patch**
These 9796 blocks translates into a dict of 9796 elements. 
PSS increased around 80MB after calling the query
RucioInjectorPoller took around 15-25 secs to execute

Comment (todor-ivanov) [reply src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py:61]: Thanks @germanfgv for doing the check. This completely overlaps with what I was expecting and was mentioning above. To me it seems pretty affordable increase of resource consumption, especially when it comes to memory. And moreover we are actually performing better in terms of iterating trough the data structure in the component (outside the DAO), as seen from your tests, because the data is already pre-aggregated. 
@vkuznet  I hope these numbers are good enough for you as well.

Comment (vkuznet) [reply src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py:61]: thanks @germanfgv and @todor-ivanov, this is certainly a very assuring observation and I'm fine to proceed with a patch.

Comment (todor-ivanov) [reply src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py:61]: Thanks @vkuznet! I'll proceed with merging now.

**Comment by amaltaro** [comment]
@germanfgv @todor-ivanov thanks for following this up to the end. In the future, please keep in mind that:
* the label saying it's "work in progress" or "do not merge yet" should be removed by the developer, to flag that it's ready to go
* commits must be squashed not to pollute the file history with unneeded commits.
Thanks!

**Comment by todor-ivanov** [comment]
Hi @amaltaro, 

This  must have been my fault. We agreed with @germanfgv for this PR he should skip the squash step because no unittest related commits were needed, but instead I was supposed to push the `Squash and Merge` button and let github do the job for us... Well, IIRC I did so, but obviously something did not go as expected.

**Comment by amaltaro** [comment]
Oh, looking at the master commit history, it indeed did the job and there is only 1 commit (I had looked only at the PR itself).

**Comment by todor-ivanov** [comment]
This actually is a good lesson not to touch this button any more (I will put some red tape on my screen right where it is :) ) 
I was actually expecting to see the `squash && merge` action to be performed also in the PR. I do consider having a single commit in the PR to be more readable for future reference. Next time we should squash the old fashion way.
